File: scripts/taskupload_server.py
#!/usr/bin/env python
import rospy
from tsu200.srv import TaskUpload, TaskUploadResponse
from geometry_msgs.msg import Point


def handle_taskupload(req):
    rospy.loginfo("TaskUpload received: id=%d, HomePos=(%.3f,%.3f,%.3f), PosNum=%d", req.id, req.HomePos.x, req.HomePos.y, req.HomePos.z, req.PosNum)
    for i, p in enumerate(req.PosList):
        rospy.loginfo("  Pos %d: (%.3f,%.3f,%.3f) type=%s info=%s", i, p.x, p.y, p.z, p.task_type, p.info)

    # Basic validation
    if req.PosNum != len(req.PosList):
        msg = 'PosNum does not match length of PosList (%d vs %d)' % (req.PosNum, len(req.PosList))
        rospy.logwarn(msg)
        return TaskUploadResponse(req.id, False, 2, msg)

    # Simulate storing/processing the uploaded task list
    try:
        # (Insert real logic here: save to DB, dispatch tasks, etc.)
        rospy.loginfo('Processing TaskUpload id=%d: %d positions', req.id, req.PosNum)
        # For example, always succeed
        return TaskUploadResponse(req.id, True, 0, 'Uploaded and scheduled')
    except Exception as e:
        rospy.logerr('Error processing TaskUpload: %s', e)
        return TaskUploadResponse(req.id, False, -1, str(e))


def taskupload_server():
    rospy.init_node('taskupload_server')
    s = rospy.Service('task_upload', TaskUpload, handle_taskupload)
    rospy.loginfo('task_upload service ready')
    rospy.spin()


if __name__ == '__main__':
    taskupload_server()

File: src/taskupload_server.cpp
#include <ros/ros.h>
#include "tsu200/TaskUpload.h"
#include "tsu200/Pos.h"
#include <geometry_msgs/Point.h>

bool handle_taskupload(tsu200::TaskUpload::Request &req, tsu200::TaskUpload::Response &res) {
  ROS_INFO("TaskUpload request id=%d PosNum=%d", req.id, req.PosNum);
  if (req.PosNum != (int)req.PosList.size()) {
    res.id = req.id;
    res.success = false;
    res.status_code = 2;
    res.message = "PosNum does not match PosList length";
    ROS_WARN("%s", res.message.c_str());
    return true;
  }
  // Process positions (simulate)
  for (size_t i = 0; i < req.PosList.size(); ++i) {
    const tsu200::Pos &p = req.PosList[i];
    ROS_INFO("  Pos %lu: x=%f y=%f z=%f type=%s info=%s", i, p.x, p.y, p.z, p.task_type.c_str(), p.info.c_str());
  }
  res.id = req.id;
  res.success = true;
  res.status_code = 0;
  res.message = "Uploaded and scheduled";
  return true;
}

int main(int argc, char **argv) {
  ros::init(argc, argv, "taskupload_server_cpp");
  ros::NodeHandle nh;
  ros::ServiceServer service = nh.advertiseService("task_upload", handle_taskupload);
  ROS_INFO("task_upload service ready (C++)");
  ros::spin();
  return 0;
}
